README.md
# Adaptive Autosar repository

This is the MPU Adaptive AUTOSAR repository for the Gen20x.i3 project.

Apart from this document, you will find more documentation under `docs/`.

## Setting up

Regardless of which of the following setup methods you use, please make sure you set up your SWF gitlab according to
[this guide](https://wiki.swf.daimler.com/display/swf/How+to+use+HTTPS+to+checkout+from+GitLab). If you don't have the
client certificate files anymore, you can re-request them
[here](https://gsep.app.corpintra.net/helpdesk/faces/portlets/index.xhtml). Our scripts will assume that you have
stored your x509 client certificates in `~/.ssh/gsep`. Our development environment scripts and automation are currently
not configured to work with SSH key authentication or `ssh-agent`. If you want to contribute this improvement, it is
welcome. Once you have done this setup, you will still use the SSH clone URLs from the Gitlab web interface, but git
will internally convert these into HTTPS requests using the OAuth token and x509 client certificates.

This document describes three ways how to build and run. The first method is preferred, but all are supported. These
methods are:

* build with devcontainers, described [below](#build-with-build_env-devcontainer)
* [setup to build with plain docker](docs/setup_without_devcontainers.md#setup-to-build-with-plain-docker) instead of
devcontainers
* [setup to build without any docker](docs/setup_without_devcontainers.md#setup-to-build-without-docker) at all

### The use of docker in this repository

Currently we use docker to build applications. It is optional, but it makes certain things easier and they enable a few
extra use-cases. For example, for building, it allows us to deliver all the dependencies and tools under one package
(NVIDIA Drive SDK, toolchains, Bazel dependencies) and have little interference from Daimler IT. In the past we also
used to use docker to run as well, but began to hit issues with network and firewall configurations, as well as not
being able to run our stack in the CI pipelines.

You can find a guide on setting up docker on Daimler Ubuntu laptops
[here](https://wiki.swf.daimler.com/display/swf/How+to+setup+docker).

We call the docker container where we build things `build_env` and sometimes `devcontainer`.

The CI environment can be found [here](devops/docker/Dockerfile.build_env). The CI/CD pipeline for this repository
pushes new version of the repository to
[SWF Artifactory](https://artifact.swf.daimler.com/list/adasdai-docker/minerva_mpu_docker/minerva_mpu/) on merge. The
image includes all the dependencies needed, including Bazel, Bazel plugins and toolchains (for both x86 and aarch64).

### Setup credentials for SWF Artifactory Docker registry

Our main docker container registry resides in SWF artifactory. To be able to download an image, first you should log
in to SWF Artifactory with your credentials:

```
docker login artifact.swf.daimler.com
```

### Build with `build_env` devcontainer

There are multiple tools which support devcontainers:
[VS Code with the Remote Development plugin](https://code.visualstudio.com/docs/remote/containers#_getting-started)
 (preferred) or the [devc CLI tool](https://github.com/nikaro/devc).

We will only be discussing the VS Code method for now. Make sure you have the
[Remote Development plugin](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack)
installed. Once this is done, open the repository with VS Code by running this command:

```
cd adaptive_autosar
code .
```

Once VS Code has started, press `Ctrl-Shift-P` and type "Focus on Containers View". Then in the top bar you will see a
`+` sign. Click on it and then select "Open Current Folder in Container View".

The first start is expected to take a long time. This is because it will fetch the base image from Artifactory, which
is currently around 25 GB. It might require your computer turned on overnight in order to finish the download.
Subsequent starts will be much quicker.

The devcontainer will set up your `git` and `ssh` inside the container to work with SWF git. This assumes you have the
environment outside the docker container already working with SWF git and that you have your x509 SWF certificates
stored in `~/.ssh/gsep`. The devcontainer is configured to run in host network mode and the proxy is configured for
you. `ssh-agent` and `gpg-agent` forwarding is handled through VS Code. Bash history and Bazel build cache is persisted
between sessions. Your user ID inside the devcontainer will be made to match the one outside the container.

## Build

The actual build instructions are the same, regardless if you use the `build_env` to build or not.

The build will typically require more than 40-50 GB of memory because it has a high degree of parallelism. If you don't
have that much phyiscal RAM on your machine, follow
[this guide](docs/bazel_tips_and_tricks.md#increasing-the-size-of-the-swap) to increase the size of the swap.

## The actual build steps

The current Bazel build is based on [rules_foreign_cc](https://github.com/bazelbuild/rules_foreign_cc) for building
external CMake projects. In particular, they are used to build the Vector BSW libraries. The default build type for
the BSW modules is "Release".

To proceed with your build on host, change to your repository root directory and execute the following commands:

```
bazel build //deployment/apricot_adaptive:filesystem_tar --config=<CONFIGURATION>
```

where `<CONFIGURATION>` is the target toolchain configuration, e.g. (`x86_64_linux`, `aarch64_linux_ubuntu` or
`aarch64_linux_linaro`, `x86_64_qnx`, `aarch64_qnx`). For now, you cannot skip `--config=x86_64_linux` even if you are
building on an `x86_64` host for an `x86_64` target. Skipping would lead to linking issue. This will be fixed in the
future by adding custom toolchain files for `x86_64_linux`.

For the `x86_64_qnx` and `aarch64_qnx` toolchains to work, you need to have the `QNX_HOST` and `QNX_TARGET` set
correctly in your environment. This is usually done by running something like: `source ~/qnx/qnx700/qnxsdp-env.sh`. If
these variables are not set, then Bazel will assume the following values:

- `QNX_HOST=/opt/qnx/qnx700/host/linux/x86_64`
- `QNX_TARGET=/opt/qnx/qnx700/target/qnx7`

If you want to add the measurement techonology tools to the stack, call bazel with `--//:mt`.

If you want to generate a package for software update build the
`//deployment/apicot_adaptive:filesystem_update_package` target. For now, generating a software update package
requires `--config=stripped`. This limitation will be removed in the future.

If you want to generate a package with only basic configuration, build the
`//deployment/apricot_adaptive:filesystem_tar --config=<CONFIGURATION>`.
## Running

### Running a virtual ECU with QEMU

This method is currently known to only work with the `build_env`. This is because it needs the NVIDIA DRIVE OS Linux
files which come as part of the `build_env` container. If you would want to make this work without the
`build_env`, you would have to make sure you have the correct DRIVE OS SDK dependencies placed at `/drive`.

Our QEMU execution environment currently supports aarch64 executables which uses Drive OS Linux  as a base, which
NVIDIA provides only for aarch64.

Alternatively, x86_64 QEMU enviroment can be brought up for tasks that are not dependable
on Nvidia Drive OS and benifit for native virtualization and KVM accleration.

The execution is currently done through a script, although this will be moved back to be done through bazel in the
future.

First, make sure you built the stack using the steps above. Use the `aarch64_linux_linaro` config. This uses the DRIVE
SDK toolchain.

Then run like so:

For aarch64 build -
```
cd os/linux # Important!
./run_aarch64_qemu.sh

```

On first run, the script will download and build the Linux kernel and it will also put together the DRIVE OS
Linux filesystem. Depending on your machine, this might be enough time for you to have lunch, a cake and a couple of
coffees. Luckily, subsequent runs will re-use the results from the first run, so you will only have to do this once.

If you open up the `os/linux/run_aarch64_qemu.sh` script, you will see a bunch of configurable parameters. The most
important is `BOOT_ADAPTIVE_STACK_TO_FOREGROUND`. When it is set to true, this will cause the QEMU instance to bind the
the Adaptive AUTOSAR stack to the main tty as soon as systemd has finished booting the system. When it is set to false,
this will cause the QEMU instance when finished bootin to drop to a login prompt and boot the Adaptive stack in the
background. The username is `nvidia` and password `nvidia`.

### Running x86_64 virtual ECU with KVM acceleration and QEMU

Alternatively, x86_64 QEMU enviroment can be brought up for tasks that are not dependable on Nvidia Drive OS and benifit
from native virtualization and KVM accleration.

For x64_64 build -
```
cd os/linux # Important!
./run_x86_64_qemu.sh

```
On first run, the script will download Ubuntu cloud image and then bazel x86_64 build for Adaptive stack is extracted into it.

Once Image boots up, you can login to the image by below credentials -
```
ubuntu login: ubuntu
Password: ubuntu

```

Alternatively, you can also ssh into the image from host with same password by below command -

```
ssh -p 10022 ubuntu@localhost

```

Adaptive stack is started in the background, check with below command -
```
sudo systemctl status adaptive-stack.service
```

### Connecting to remote DLT

All of the applications apart from the `log-daemon` are configured to output remote DLT. The connection is available
over TCP at IP address `10.21.17.98` and port `49361`.

When running a virtual ECU inside QEMU, the QEMU configuration will forward the remote DLT port to port `13490` on the
host machine. Configure the [dlt-viewer](https://github.com/GENIVI/dlt-viewer) running on your host to connect to
`127.0.0.1` port `13490` over TCP.

## Releasing

To release a package to Artifactory, use the following
[pipeline](https://sfci.swf.daimler.com/job/apricotbscqal/view/Component/job/component.adaptive-autosar.master.manual/).

You need to specify a `RELEASE_LABEL` before starting the job. This parameter is the tag/version of the new release.
A good value would be an increment of the previous release.

Releases are stored
[here](https://artifact.swf.daimler.com/webapp/#/artifacts/browse/tree/General/apricot-bin-delivery/xpf/adaptive-autosar).

## Miscellaneous

You can use ssh to connect to the virtual ECU.
To connect to Qemu VM instance over ssh from host, we forward host port 10022 to guest port 22.
More information on how to set this up is [here](https://git.swf.daimler.com/adasdai/minerva_mpu_adaptive/-/blob/master/docs/setup_ssh.md)

### Build collectd and its plugins through bazel

collectd can be built and install binaries using:
```
bazel build //tools:build_collectd
```
The binaries are generated in `bazel-bin/tools` directory.

## Contributing

We support a usual merge request-based contribution flow. For now, you don't have to push the your proposal as a branch
in a personal gitlab fork. You can just push it in the main repo. However, this might change. The branch naming
conventions are covered [here](https://wiki.swf.daimler.com/display/GREEN/Naming+Conventions+for+the+branches). Please
try as much as possible to link your MRs to Jira tickets.

Before you open a merge request (MR), please check the contributor checklist below.

Unless otherwise agreed, the MR threads need to be marked as resolved by the person who opened the discussion, not by
the MR author.

If the MR is not meant to be reviewed yet, or you wish to prevent it from being merged, use the "WIP:" prefix in the
title.

The preferred case for file paths, names, targets, etc. is `underscore_case`. This is mainly due to Python and bazel.

Consider the following content for your `.vscode/settings.json`:

```json
{
    "[git-commit]": {
        "editor.rulers": [
            50,
            72
        ]
    },
    "[markdown]": {
        "editor.rulers": [120]
    },
    "files.trimTrailingWhitespace": true,
    "files.insertFinalNewline": true,
}
```

### Pre-commit checks

We use a took called `pre-commit` for various static checks before creating a commit. To enable `pre-commit`,
after cloning the repository:

```
cd adaptive_autosar
pre-commit install
```

Once enabled, pre-commit will run before every local commit in order to suggest fixes for the checks defined in
[.pre-commit-config.yaml](./pre-commit-config.yaml). These checks are performed automatically in the CI pipeline.

## Contributor checklist

The following checklist is derived from a list of common mistakes.

I, the author of a merge request, have:

* checked for typos
* updated any relevant documentation files and UML diagrams
* made sure markdown paragraphs wrap at 120 characters (not automated yet)
* made sure the automated pre-commit checks pass

If I want to preserve git history, I have:

* made sure [50](https://chris.beams.io/posts/git-commit/#limit-50)/
[72](https://chris.beams.io/posts/git-commit/#wrap-72) rule for git messages is respected (not automated yet)
* made sure git messages are [descriptive and well formed](https://chris.beams.io/posts/git-commit/)
* made sure the automated pre-commit checks pass
* made sure that if I use the apply suggestion feature in gitlab, I cleaned-up the commit log and squashed/fixed up the
modifications

Otherwise, if I want to squash the git history through gitlab, I have:

* made sure [50](https://chris.beams.io/posts/git-commit/#limit-50)/
[72](https://chris.beams.io/posts/git-commit/#wrap-72) rule for the gitlab MR title and description is respected (not
automated yet)
* made sure the MR title and description is [descriptive and well formed](https://chris.beams.io/posts/git-commit/)


CHANGELOG.md
# Change Log

## Unreleased

### Changed

* Release step 1B5.1 of AMSR is used.
* The IDC ECU extract replaces the Starter Kit ECU extract that was used until now.
* Applications each now have its own `package` target implemented using Bazel `pkg_tar`. The application and the config
files get built into the filesystem using this package.
* The measurement technology is disabled by default for BROP B2. It is now included in the filesystem only if you call
`bazel build` with `--//:mt`.
* Implemented a cyclic dependency workaround for socal which doesn't require extra targets and extra commands to be
run.
* An application base library has been created to contain the common parts for Adaptive applications. Applications
should add `//application/common:application_base` to their `deps` and derive from the base class.
* In `run_aarch64_qemu.sh`, `BOOT_INTO_ADAPTIVE_STACK` was changed into `BOOT_ADAPTIVE_STACK_TO_FOREGROUND`. When this
is set to false, it will boot the stack in the background instead of not booting at all.
* The `//deployment/minerva_mpu_adaptive:filesystem_tar` target and minerva related files are permanently removed.
Only `//deployement/apricot_adaptive:filesystem_tar` target is available now.

### Deprecated

* The `//:minerva_mpu_adaptive_filesystem` is deprecated. Please use the
`//deployment/minerva_mpu_adaptive:filesystem_tar` target instead.


os/linux/run_aarch64_qemu.sh
#!/bin/bash

set -e

# Deployment (apricot_adaptive
if [ $# -eq 0 ]; then
    DEPLOYMENT="apricot_adaptive"
else 
    DEPLOYMENT=$1
fi

# If this flag is set to true the adaptive stack will bind directly to the main
# tty. If set to false, it will boot in the background.
BOOT_ADAPTIVE_STACK_TO_FOREGROUND=true

# Path to tar to include at the root of the filesystem, usually to the adaptive
# stack tar.
PATH_TO_ADAPTIVE_TAR="../../../bazel-bin/deployment/$DEPLOYMENT/filesystem_tar.tar"

if [ ! -f $PATH_TO_ADAPTIVE_TAR ]; then
    echo "File $PATH_TO_ADAPTIVE_TAR does not exist. Exiting."
    exit 0
fi

# Extra space to add on the disk. Should be big enough to fit the unpacked tar
# from above.
EXTRA_SPACE_ON_DISK="4G"

# How much memory should the emulator have.
EMULATOR_MEMORY_SIZE="1024M"

# How many cores is the guest permitted to use.
EMULATOR_SMP=4

if [ ! -d "runtime" ]; then
    mkdir runtime
fi

cd runtime

if [ ! -f "kernel-build/arch/arm64/boot/Image" ]; then
    printf "Fetch and build kernel...\n"

    wget -c -N https://cdn.kernel.org/pub/linux/kernel/v5.x/linux-5.12.tar.xz

    tar xf linux-5.12.tar.xz

    cp ../configs/apricot_qemu_defconfig linux-5.12/arch/arm64/configs

    export ARCH=arm64
    export CROSS_COMPILE=/drive/toolchains/gcc-linaro-7.3.1-2018.05-x86_64_aarch64-linux-gnu/bin/aarch64-linux-gnu-
    export LOCALVERSION="-tegra"

    mkdir kernel-build
    make -C linux-5.12 O=${PWD}/kernel-build apricot_qemu_defconfig

    # Using up all the cores might lock up your machine
    CORES_FOR_LINUX_BUILD=$(expr `nproc` - 1)

    make -j$CORES_FOR_LINUX_BUILD -C linux-5.12 O=${PWD}/kernel-build
fi

printf "Preparing filesystem...\n"

if [ ! -f "driveos.ext4.qcow2" ]; then
    virt-make-fs --format=qcow2 --type=ext4 --size=+$EXTRA_SPACE_ON_DISK /drive/drive-t186ref-linux/targetfs/ driveos-tmp.ext4.qcow2

    # Enable networking with systemd-networkd
    guestfish -a driveos-tmp.ext4.qcow2 -i ln-sf /lib/systemd/system/systemd-networkd-wait-online.service /etc/systemd/system/network-online.target.wants/systemd-networkd-wait-online.service
    guestfish -a driveos-tmp.ext4.qcow2 -i ln-sf /lib/systemd/system/systemd-networkd.socket /etc/systemd/system/sockets.target.wants/systemd-networkd.socket
    guestfish -a driveos-tmp.ext4.qcow2 -i ln-sf /lib/systemd/system/systemd-networkd.service /etc/systemd/system/multi-user.target.wants/systemd-networkd.service
    guestfish -a driveos-tmp.ext4.qcow2 -i ln-sf /lib/systemd/system/systemd-networkd.service /etc/systemd/system/dbus-org.freedesktop.network1.service

    # Activate sshd
    guestfish -a driveos-tmp.ext4.qcow2 -i rm /etc/ssh/sshd_not_to_be_run

    # Disable OEM setup for NVIDIA. We automate it.
    guestfish -a driveos-tmp.ext4.qcow2 -i ln-sf /lib/systemd/system/multi-user.target /lib/systemd/system/default.target

    # Currently the modules are built for a different kernel than we use.
    guestfish -a driveos-tmp.ext4.qcow2 -i rm-rf /lib/modules
    guestfish -a driveos-tmp.ext4.qcow2 -i rm /etc/modules-load.d/modules.conf

    # Clean-up and streamline initialization
    guestfish -a driveos-tmp.ext4.qcow2 -i rm /etc/systemd/system/multi-user.target.wants/drive-setup.service
    guestfish -a driveos-tmp.ext4.qcow2 -i rm /etc/systemd/system/multi-user.target.wants/nv_duv3.service
    guestfish -a driveos-tmp.ext4.qcow2 -i rm /etc/systemd/system/multi-user.target.wants/nv_network_configure.service
    guestfish -a driveos-tmp.ext4.qcow2 -i rm /etc/systemd/system/network-online.target.wants/nv_networkd_wait_online.service
    guestfish -a driveos-tmp.ext4.qcow2 -i rm /etc/systemd/system/ssh.service.wants/nv_ssh_host_keys.service
    guestfish -a driveos-tmp.ext4.qcow2 -i rm-rf /etc/systemd/system/tegra.target.wants

    mv driveos-tmp.ext4.qcow2 driveos.ext4.qcow2
fi

rm -f adaptive_overlay.ext4.qcow2
qemu-img create -b driveos.ext4.qcow2 -f qcow2 adaptive_overlay.ext4.qcow2

# Install systemd service for adaptive-stack
SERVICE_PATHS_TO_COPY="../configs/services/adaptive-stack.service ../configs/services/enable-ipv6-loopback.service"

if [ "$BOOT_ADAPTIVE_STACK_TO_FOREGROUND" = true ] ; then
    # If we boot to foreground, we also copy the drop-in configuration
    SERVICE_PATHS_TO_COPY="${SERVICE_PATHS_TO_COPY} ../configs/services/adaptive-stack.service.d"
fi

virt-copy-in -a adaptive_overlay.ext4.qcow2 $SERVICE_PATHS_TO_COPY /lib/systemd/system/

guestfish -a adaptive_overlay.ext4.qcow2 -i ln-sf /lib/systemd/system/adaptive-stack.service /etc/systemd/system/multi-user.target.wants/adaptive-stack.service
guestfish -a adaptive_overlay.ext4.qcow2 -i ln-sf /lib/systemd/system/enable-ipv6-loopback.service /etc/systemd/system/network-online.target.wants/enable-ipv6-loopback.service

# Configure the network interfaces
virt-copy-in -a adaptive_overlay.ext4.qcow2 ../configs/enp0s1.network /etc/systemd/network
virt-copy-in -a adaptive_overlay.ext4.qcow2 ../configs/enp0s2.network /etc/systemd/network
virt-copy-in -a adaptive_overlay.ext4.qcow2 ../configs/enp0s3.network /etc/systemd/network
virt-copy-in -a adaptive_overlay.ext4.qcow2 ../configs/enp0s4.network /etc/systemd/network
virt-copy-in -a adaptive_overlay.ext4.qcow2 ../configs/enp0s5.network /etc/systemd/network

# Add the adaptive stack to the filesystem
virt-tar-in -a adaptive_overlay.ext4.qcow2 $PATH_TO_ADAPTIVE_TAR /

# Add ssh keys
virt-copy-in -a adaptive_overlay.ext4.qcow2 ../configs/ssh/* /etc/ssh

printf "Booting...\n"

# The following command set up several virtual network interfaces for QEMU, but
# it also has the following port-forwarding to guest:
# - host port 10022 to guest port 22 on IP 10.21.17.98 for SSH
# - host port 13490 to guest port 49361 on IP 10.21.17.98 for remote DLT
# - host port 10023 to guest port 23 on IP 10.21.17.98 for telnet
# - host port 13400 to guest port 13400 on IP 169.254.17.7 for Socket Connection

qemu-system-aarch64 \
-nographic \
-machine virt \
-cpu cortex-a53 \
-smp $EMULATOR_SMP \
-m $EMULATOR_MEMORY_SIZE \
-bios /usr/share/qemu-efi-aarch64/QEMU_EFI.fd \
-kernel kernel-build/arch/arm64/boot/Image \
-append "console=ttyAMA0 root=/dev/vda rw quiet" \
-drive file=adaptive_overlay.ext4.qcow2,format=qcow2,if=none,id=drive0 \
-device virtio-blk-device,drive=drive0 \
-device virtio-net-pci,netdev=net1 \
-netdev user,id=net1,net=10.1.17.0/16 \
-device virtio-net-pci,netdev=net20 \
-netdev user,id=net20,net=10.20.17.0/16 \
-device virtio-net-pci,netdev=net21 \
-netdev user,id=net21,net=10.21.17.0/24,hostfwd=tcp::13490-10.21.17.98:49361,hostfwd=tcp::10022-10.21.17.98:22,hostfwd=tcp::10023-10.21.17.98:23 \
-device virtio-net-pci,netdev=net127 \
-netdev user,id=net127,net=10.127.17.0/16 \
-device virtio-net-pci,netdev=net254 \
-netdev user,id=net254,net=169.254.18.0/16,hostfwd=tcp::13400-169.254.17.7:13400
